components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/device/UserAuthenticationEndpoint.java

```diff
@@ -98,48 +108,91 @@ public Response deviceAuthorize(@Context HttpServletRequest request, @Context Ht
                     commonAuthRequestWrapper.setParameter(Constants.SCOPE, scope);
                 }
                 commonAuthRequestWrapper.setParameter(Constants.NONCE, userCode);
+                // Set the client authentication context to the request for API based authentication flow.
+                if (OAuth2Util.isApiBasedAuthenticationFlow(request)) {
+                    setClientAuthnContext(request, clientId);
+                }
                 return oAuth2AuthzEndpoint.authorize(commonAuthRequestWrapper, response);
             } else {
                 if (log.isDebugEnabled()) {
                     log.debug("Incorrect user_code.");
                 }
+                if (OAuth2Util.isApiBasedAuthenticationFlow(request)) {
+                    return handleApiBasedAuthenticationErrorResponse(HttpServletResponse.SC_BAD_REQUEST,
+                            Error.INVALID_REQUEST, INVALID_CODE_ERROR_KEY, "user_code is invalid or expired.");
+                }
                 String error = ServiceURLBuilder.create().addPath(Constants.DEVICE_ENDPOINT_PATH)
                         .addParameter(ERROR, INVALID_CODE_ERROR_KEY).build().getAbsolutePublicURL();
                 return Response.status(HttpServletResponse.SC_FOUND).location(URI.create(error)).build();
             }
         } catch (IdentityOAuth2Exception e) {
-            return handleIdentityOAuth2Exception(e);
+            return handleIdentityOAuth2Exception(e, OAuth2Util.isApiBasedAuthenticationFlow(request));
         } catch (URLBuilderException e) {
-            return handleURLBuilderException(e);
+            return handleURLBuilderException(e, OAuth2Util.isApiBasedAuthenticationFlow(request));
         } catch (URISyntaxException e) {
-            return handleURISyntaxException(e);
+            return handleURISyntaxException(e, OAuth2Util.isApiBasedAuthenticationFlow(request));
         }
     }
 
-    private Response handleIdentityOAuth2Exception(IdentityOAuth2Exception e) throws OAuthSystemException {
+    private void setClientAuthnContext(HttpServletRequest request, String clientId) {
```

Comment:
for device code if app native authentication enabled, client doesn't need to authenticate it self is that so?

Reply:
For device code grant, clientID is not sent in the request. ClientID is retrieved using the user_code